### ISSUE TITLE
Rikaichamp can reposition the text cursor

### ISSUE THREAD

**Issue by kuchitsu** [birtles/rikaichamp#18]
While Rikaichamp is enabled, I can reposition the text cursor by pointing at different parts of the text. I assume this isn't intended behavior.

This is what I'm talking about:
https://i.imgur.com/r89LGnF.png
https://i.imgur.com/cg74Pfb.png

**Comment by birtles** [comment]
I guess that text is in a text box?

In web browsers there is an API for selections but it doesn't apply to text boxes (text areas and text inputs): they have their own separate selection API. Text boxes always have a selection even if it's just the text cursor. In order for rikaichamp to highlight text in textboxes, it needs to save the original selection and then restore it after it is done. It normally restores the selection when you mouse over something else. In this case, perhaps it is restoring the selection incorrectly.

Do you have a reliable set of steps to reproduce the problem?

**Comment by kuchitsu** [comment]
Yes, it's in a text box.

Reproducing it is very trivial on my Firefox. Write any text, point at it somewhere, done. Here is a [simple demonstration](https://gfycat.com/gifs/detail/RelievedThatGangesdolphin). See how it's jumping all over the place?

**Comment by birtles** [comment]
Yes, however if you restore the selection as soon as the popup disappears and you've scrolled (or selecting has caused the browser to scroll the text box) then suddenly the scroll position will jump back to show the original selection. So we'd probably need to record the current scroll position, restore the selection, then update the scroll position again.

**Comment by alexanderabramov** [comment]
As a workaround until this issue can be addressed, it is possible to disable "Highlight matched text" in the options.

**Comment by birtles** [comment]
Sorry for the delay here. I hope to get to this later this week.

**Comment by WGH-** [comment]
It selects text inside input boxes when cursor points just somewhere inside the box, not even directly at the text. It make editing very annoying, e.g I place a cursor inside the box, start typing, and then Rikaichamp popup suddenly appears, text becomes selected, and my next keypress overwrites all the text.

https://wgh.torlan.ru/ss/1514839830-Mf.ogv

Is it possible to highlight the text without using "browser built-in highlight" that interferes with clipboard and cursors?

**Comment by birtles** [comment]
Yes, I appreciate the frustration. This is next on my list of bugs to fix. I'm optimistic that recording and restoring the scroll position will work but I'll check that latest test case too.

Unfortunately the alternative to using the browser's selection API is probably synthesizing an overlay element that has, as fas as possible, identical styling to the underlying textbox, and doing selection on that instead. That would work for a lot of cases and avoid touching the text box contents, however I expect that in some cases the overlay element would not match the underlying style exactly, or would not be removed at the right time, and would actually cause more problems.

**Comment by WGH-** [comment]
I wonder how the original RikaiChan does this. I don't remember having such problems.

**Comment by birtles** [comment]
I started working and I think I fixed the initial problem. However, there's still seems to be a bug where, if you however over some part of the textbox that doesn't have text it will jump back to selecting the start of the text box (since that's what `caretPositionFromPoint` is going to give us in that case).

I think we probably need to add something similar to #11 but for text boxes. Perhaps it could be a lot simpler, however, and, if the returned position is zero then just check that the cursor is, in fact, pretty close to the zero position (which will depend in the `writing-mode`). It will also need to be careful to check the scroll position too.
